src/cpp/include/openvino/genai/llm_pipeline.hpp

```diff
@@ -0,0 +1,216 @@
+        return generate(text, AnyMap{std::forward<Properties>(properties)...});
+    }
+    std::string generate(std::string text, const ov::AnyMap& config);
+
+    template <typename... Properties>
+    util::EnableIfAllStringAny<EncodedResults, Properties...> generate(
+        ov::Tensor input_ids,
+        Properties&&... properties) {
+        return generate(input_ids, AnyMap{std::forward<Properties>(properties)...});
+    }
+    EncodedResults generate(ov::Tensor input_ids, const ov::AnyMap& config);
+
+    /**
+    * @brief High level generate for batched prompts which encodes inputs and returns decoded outputs. 
+    * Streamer cannot be used for multibatch inputs.
+    *
+    * @param text input prompt
+    * @param generation_config optional GenerationConfig
+    * @return DecodedResults a structure with resulting texts & scores
+    */
+    DecodedResults generate(const std::vector<std::string>& texts, OptionalGenerationConfig generation_config);
+
+    /**
+    * @brief Low level generate to be called with already encoded input_ids tokens.
+    * Streamer cannot be used for multibatch inputs.
+    *
+    * @param input_ids encoded input prompt tokens
+    * @param attention_mask optional attention_mask
+    * @param generation_config optional GenerationConfig
+    * @param streamer optional streamer
+    * @return EncodedResults a structure with resulting tokens and scores
+    * @throws Exception if the stremaer is set for inputs_ids with multiple batches
+    */
+    EncodedResults generate(ov::Tensor input_ids, 
+                            std::optional<ov::Tensor> attention_mask, 
+                            OptionalGenerationConfig generation_config=std::nullopt,
+                            OptionalStreamerVariant streamer=std::nullopt);
+    
+    template <typename InputsType, typename... Properties>
+    util::EnableIfAllStringAny<std::string, Properties...> operator()(
+        InputsType text,
+        Properties&&... properties) {
+        return generate(text, AnyMap{std::forward<Properties>(properties)...});
+    }
+    
+    DecodedResults operator()(const std::vector<std::string>& text, OptionalGenerationConfig generation_config=std::nullopt) {
+        return generate(text, generation_config);
+    }
+
+    std::string operator()(
+        std::string text, 
+        OptionalGenerationConfig generation_config=std::nullopt, 
+        OptionalStreamerVariant streamer=std::nullopt
+    ) {
+        return generate(text, generation_config, streamer);
+    }
+    
+    ov::genai::Tokenizer get_tokenizer();
+    GenerationConfig get_generation_config() const;
+    void set_generation_config(const GenerationConfig& generation_config);
+
+    void start_chat();
+    void finish_chat();
+    void reset_state();
+    std::string apply_chat_template(std::string prompt, std::string role = "user") const;
+private:
+    class LLMPipelineImpl;
+    std::unique_ptr<LLMPipelineImpl> m_pimpl;
```

Comment:
@ilya-lavrenov Can it be an interface instead? NPU-friendly LLM execution approach differs from other plugins (CPU/GPU), so it could be isolated into separate implementation.

Reply:
NPU / GPU or other device specific will be hidden inside `src` folder and will not be reflected in API

Reply:
out of curiosity, What's the difference for NPU?

Reply:
Yes, it shouldn't be part of public API for sure.

Now [LLMPipelineImpl](https://github.com/openvinotoolkit/openvino.genai/blob/9e37273e2528d948c655b0710e3d504d1a7d82c1/src/cpp/src/llm_pipeline.cpp#L84) is the concrete type (not the interface) and all the NPU specific details will probably be hidden somewhere there which could make logic a little complicated.

My point was, if it could be an interface `ILLMPipelineImpl` we could introduce NPU-friendly pipeline as different implementation and don't mix it with existing one. E.g:
```
ov::genai::LLMPipeline::LLMPipeline(
    const std::string& model_path,
    const ov::genai::Tokenizer& tokenizer,
    const std::string& device,
    const ov::AnyMap& plugin_config
) {
    if (device == "NPU") {
        m_pimpl = make_unique<LLMNPUPipeline>(model_path, tokenizer, device, plugin_config)
    } else {
        m_pimpl = make_unique<LLMPipelineImpl>(model_path, tokenizer, device, plugin_config);
    }
}
```
Running LLM on NPU will require two static shapes models (prefill, generation) with KV-cache exposed as input and outputs. So it will be different from what you have for CPU / GPU plugins.

Reply:
`LLMNPUPipelineImpl` contains a lot of stuff and NPU will have to duplicate a lot. 
Probably, we need to integrate NPU on even lower level

Reply:
That's ok, probably we should discuss it separately, thanks!